### PR TITLE
[debugger] Changing how debugger handles exception when hybrid suspend is enabled

### LINKED ISSUE #15203: Debugger doesn't break on unhandled exceptions
## Steps to Reproduce

1. Write simple console application throwing unhandled exception
```
using System;

namespace ConsoleApplication1
{
    internal class Program
    {
        public static void Main(string[] args)
        {
            throw new ArgumentException("test");
        }
    }
}
```
2. Run debug: `mono --debug --debugger-agent=transport=dt_socket,server=y,address=127.0.0.1:55555 Program.exe`

### Current Behavior

No `UnhandledException` event sent, so a debugger doesn't break on unhandled exceptions. 

### Expected Behavior

A debugger breaks on unhandled exceptions

## On which platforms did you notice this

[x] macOS
[x] Linux
[ ] Windows

**Mono 5.20.1.19**:
```
Mono JIT compiler version 5.20.1.19 (2018-10/886c4901747 Tue Apr  9 12:37:29 EDT 2019)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
        TLS:           
        SIGSEGV:       altstack
        Notification:  kqueue
        Architecture:  amd64
        Disabled:      none
        Misc:          softdebug 
        Interpreter:   yes
        LLVM:          yes(600)
        Suspend:       hybrid
        GC:            sgen (concurrent by default)
```

Also reproduces on **Mono 6.0.0.299**.
Works as expected when using **Mono 5.18.1.3**.

I tried it with JetBrains Rider and VS Code and both don't break on unhandled exceptions.

I also tried to reproduce the scenario on Windows with Mono **5.20.1.19** and everything works as expected.

### LINKED ISSUE #17084: Marshalling managed exception contains cut off stacktraces
_From @Therzok on Thu, 26 Sep 2019 15:12:34 GMT_

### Steps to Reproduce

1. Run the sample zip
2. Click on the window.

### Expected Behavior
There should be an exception logged by the application
```
System.Exception: e
  at bubbleexception.ViewController.MouseDown () [0x00009] in /Users/therzok/Projects/bubbleexception/bubbleexception/ViewController.cs:25 
--- End of stack trace from previous location where exception was thrown ---

  at (wrapper managed-to-native) AppKit.NSApplication.NSApplicationMain(int,string[])
  at AppKit.NSApplication.Main (System.String[] args) [0x00040] in /Library/Frameworks/Xamarin.Mac.framework/Versions/6.2.0.42/src/Xamarin.Mac/AppKit/NSApplication.cs:100 
  at bubbleexception.MainClass.Main (System.String[] args) [0x00005] in /Users/therzok/Projects/bubbleexception/bubbleexception/Main.cs:14 
```

### Actual Behavior
The exception is only partly shown, seemingly only for the inner exception

```
System.Exception: e
  at (wrapper managed-to-native) AppKit.NSApplication.NSApplicationMain(int,string[])
  at AppKit.NSApplication.Main (System.String[] args) [0x00040] in /Library/Frameworks/Xamarin.Mac.framework/Versions/6.2.0.42/src/Xamarin.Mac/AppKit/NSApplication.cs:100 
  at bubbleexception.MainClass.Main (System.String[] args) [0x00008] in /Users/therzok/Projects/bubbleexception/bubbleexception/Main.cs:14 
```

### Environment

<!--
1.
```

Visual Studio: Help > About Microsoft Visual Studio > Copy Info [button]
Visual Studio for Mac: Visual Studio > About Visual Studio > Show Details > Copy Information [button]
2. Paste into the code block below (between ```)
-->

```
=== Visual Studio Enterprise 2019 for Mac ===

Version 8.3.1 (build 6)
Installation UUID: 010a6a17-65d7-44f0-9e3d-48e5afe82e81
	GTK+ 2.24.23 (Raleigh theme)
	Xamarin.Mac 5.16.1.24 (d16-3 / 08809f5b)

	Package version: 606000082

=== Mono Framework MDK ===

Runtime:
	Mono 6.6.0.82 (2019-08/70d69030534) (64-bit)
	Package version: 606000082

=== NuGet ===

Version: 5.3.0.6192

=== .NET Core SDK ===

SDK: /usr/local/share/dotnet/sdk/3.0.100/Sdks
SDK Versions:
	3.0.100
	3.0.100-rc1-014190
	3.0.100-preview9-014004
	3.0.100-preview8-013656
	3.0.100-preview6-012264
	2.2.203
	2.1.701
	2.1.700
	2.1.603
MSBuild SDKs: /Library/Frameworks/Mono.framework/Versions/6.6.0/lib/mono/msbuild/Current/bin/Sdks

=== .NET Core Runtime ===

Runtime: /usr/local/share/dotnet/dotnet
Runtime Versions:
	3.0.0
	3.0.0-rc1-19456-20
	3.0.0-preview9-19423-09
	3.0.0-preview6-27804-01
	2.2.7
	2.2.4
	2.1.12
	2.1.11
	2.1.10

=== Xamarin.Profiler ===

'/Applications/Xamarin Profiler.app' not found

=== Updater ===

Version: 11

=== Apple Developer Tools ===

Xcode 11.0 (14936)
Build 11A420a

=== Xamarin.Mac ===

Version: 6.2.0.42 (Visual Studio Enterprise)
Hash: 5e8a208b
Branch: d16-3
Build date: 2019-09-17 14:19:30-0400

=== Xamarin.iOS ===

Version: 13.2.0.42 (Visual Studio Enterprise)
Hash: 5e8a208b
Branch: d16-3
Build date: 2019-09-17 14:19:30-0400

=== Xamarin Designer ===

Version: 16.3.0.244
Hash: e15b2ee97
Branch: remotes/origin/d16-3
Build date: 2019-09-25 00:03:16 UTC

=== Xamarin.Android ===

Version: 10.0.0.43 (Visual Studio Enterprise)
Commit: xamarin-android/d16-3/8af1ca8
Android SDK: /Users/therzok/Library/Developer/Xamarin/android-sdk-macosx
	Supported Android versions:
		4.4    (API level 19)
		4.4.87 (API level 20)
		5.0    (API level 21)
		5.1    (API level 22)
		6.0    (API level 23)
		7.0    (API level 24)
		7.1    (API level 25)
		8.0    (API level 26)
		8.1    (API level 27)

SDK Tools Version: 26.1.1
SDK Platform Tools Version: 28.0.0
SDK Build Tools Version: 27.0.0

Build Information: 
Mono: mono/mono/2019-06@7af64d1ebe9
Java.Interop: xamarin/java.interop/d16-3@5836f58
LibZipSharp: grendello/LibZipSharp/d16-3@71f4a94
LibZip: nih-at/libzip/rel-1-5-1@b95cf3f
ProGuard: xamarin/proguard/master@905836d
SQLite: xamarin/sqlite/3.27.1@8212a2d
Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-3@cb41333

=== Microsoft Mobile OpenJDK ===

Java SDK: /Users/therzok/Library/Developer/Xamarin/jdk/microsoft_dist_openjdk_1.8.0.25
1.8.0-25
Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

=== Android SDK Manager ===

Version: 1.4.0.65
Hash: c33b107
Branch: remotes/origin/d16-3
Build date: 2019-09-19 20:42:44 UTC

=== Android Device Manager ===

Version: 1.2.0.115
Hash: 724ea69
Branch: remotes/origin/d16-3
Build date: 2019-09-19 20:43:06 UTC

=== Xamarin Inspector ===

Version: 1.4.3
Hash: db27525
Branch: 1.4-release
Build date: Mon, 09 Jul 2018 21:20:18 GMT
Client compatibility: 1

=== Build Information ===

Release ID: 803010006
Git revision: e4b65219fee9cb778751f3bae5b305111c51c46b
Build date: 2019-09-25 18:56:10+00
Build branch: release-8.3
Xamarin extensions: 17def12334dfec3725cc3b26d3f79b621182e4a1

=== Operating System ===

Mac OS X 10.14.6
Darwin 18.7.0 Darwin Kernel Version 18.7.0
    Tue Aug 20 16:57:14 PDT 2019
    root:xnu-4903.271.2~2/RELEASE_X86_64 x86_64

=== Enabled user installed extensions ===

Log Monitor 0.2

```

### Build Logs

<!--
1. Place cursor below this comment block.
2. Attach build log or link to gist (https://gist.github.com/) of the log.
3. Logs can be found in the Build Output tab of the Errors pad in Visual Studio
To get full build logs, set the log verbosity to diagnostic at the following locations:
- On Visual Studio for Mac: Preferences > Projects > Build
- On Visual Studio for Windows: Tools > Options > Projects and Solutions > Build and Run
On Visual Studio for Windows you also want to add `-v -v -v -v` to the mtouch additional arguments by right-clicking the project in the solution explorer and selecting `Properties`.
-->

### Example Project (If Possible)

<!--
1. Place cursor below this comment block.
2. Drag and drop the compressed project or files needed to reproduce.
-->

<!--
Switch to the "Preview" tab to ensure your issue renders correctly.
-->
[bubbleexception.zip](https://github.com/xamarin/xamarin-macios/files/3658472/bubbleexception.zip)


_Copied from original issue xamarin/xamarin-macios#7117_

### PR DESCRIPTION
**Pull request by thaystg**
Removing all fixes to fix #15203 and #17083, and trying a new approach to fix this #17084, together.

The thing is when we are executing on hybrid suspend the caller is executed inside try catch in the runtime_invoke_ method that is generated. And this try catch should be ignored for the debugger exception, otherwise it's the last try catch in main thread, but should always be considered to generate the stack trace.

Fixes #17084

### CODE REVIEW

**Comment by marek-safar** [comment]
@vargaz could you please review

File: mono/mini/mini-exceptions.c [hunk 1]
```diff
@@ -2555,6 +2555,8 @@ mono_handle_exception_internal (MonoContext *ctx, MonoObject *obj, gboolean resu
 	MonoObject *non_exception = NULL;
 	Unwinder unwinder;
 	gboolean in_interp;
+	gboolean is_catched_unmanaged = FALSE;
```
Comment (BrzVlad) [inline mono/mini/mini-exceptions.c]: caught :D

**Comment by thaystg** [comment]
@monojenkins build failed

**Comment by thaystg** [comment]
@monojenkins backport 2019-08

**Comment by thaystg** [comment]
@marek-safar , should we backport to 2019-06?

**Comment by marek-safar** [comment]
> should we backport to 2019-06?

@thaystg: no

**Pull request merged**